### PR TITLE
Addition of ReLU6

### PR DESCRIPTION
**Pull request by Aakash-kaushik**
Initial draft PR for addition of ReLU6
PyTorch reference [here](https://pytorch.org/docs/stable/generated/torch.nn.ReLU6.html)
Requirement: Needed to Implement [Mobilenet V1](https://github.com/mlpack/models/pull/72)

Function Originally introduced in the paper: [MobileNets: Efficient Convolutional Neural Networks for Mobile Vision Applications](https://arxiv.org/pdf/1704.04861)

### CODE REVIEW

File: src/mlpack/methods/ann/layer/relu6_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void ReLu6<InputDataType, OutputDataType>::Forward(
+    const InputType& input, OutputType& output)
+{
+  output.zeros();
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/relu6_impl.hpp:43]: That only works if output is already initialized, if not you have to use `output = arma::zeros<OutputDataType>(arma::size(input));`

**Comment by Aakash-kaushik** [comment]
Values for tests generated from the below pytorch code 
```python3
import torch
import torch.nn as nn

relu = nn.ReLU6()
inp = torch.tensor([-2.0, 3, 0, 6, 24] ,requires_grad=True)
o = relu(inp)
print(inp)
print(o)
o.sum().backward()
print(inp.grad)
```
output:
```bash
tensor([-2.,  3.,  0.,  6., 24.], requires_grad=True)
tensor([0., 3., 0., 6., 6.], grad_fn=<HardtanhBackward0>)
tensor([0., 1., 0., 0., 0.])
```
Also the PR is ready for review.

**Comment by zoq** [comment]
> And I wanted to ask if a gradient function is necessary for a relu activation function? I am a bit clueless about that and if that's not the case the PR is ready for a review.

You only have to implement the `Gradient` function if the function has a learnable parameter, which isn't the case for `relu6`.

File: src/mlpack/methods/ann/layer/relu6.hpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  template<typename DataType>
+  void Backward(const DataType& input, const DataType& gy, DataType& g);
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get size of weights.
+  size_t WeightSize() const { return 0; }
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/relu6.hpp:76]: Some function are missing from the layer, like `Delta()`.

File: src/mlpack/methods/ann/layer/relu6_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename DataType>
+void ReLU6<InputDataType, OutputDataType>::Backward(
+    const DataType& input, const DataType& gy, DataType& g)
+{
+  DataType derivative(arma::size(gy));
+  derivative.fill(0.0);
+  for (size_t i=0; i < input.n_elem; ++i)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/relu6_impl.hpp:56]:
```suggestion
  for (size_t i = 0; i < input.n_elem; ++i)
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 4]
```diff
@@ -704,6 +704,52 @@ void CheckFlattenTSwishDerivateCorrect(const arma::colvec input,
...
+TEST_CASE("ReLU6FunctionTest", "[ActivationFunctionsTest]")
+{
+  const arma::colvec activationData("-2.0 3.0 0.0 6.0 24.0");
+
+  const arma::colvec desiredActivations("0.0 3.0 0.0 6.0 6.0");
+
+  const arma::colvec desiredDerivatives("0.0 1.0 0.0 0.0 0.0");
+
+  CheckReLU6Correct(activationData, desiredActivations, desiredDerivatives);
+
```
Comment (zoq) [inline src/mlpack/tests/activation_functions_test.cpp:750]:
```suggestion
```

**Comment by Aakash-kaushik** [comment]
Will make the changes by tomorrow morning, thanks for the review.

**Comment by Aakash-kaushik** [comment]
> Will make the changes by tomorrow morning, thanks for the review.

@zoq Made the changes.

**Comment by kartikdutt18** [comment]
@zoq, Correct me if I'm wrong, but I think this should be part of activation function(https://github.com/mlpack/mlpack/tree/master/src/mlpack/methods/ann/activation_functions) not layers since it doesn't require a parameter.

**Review by kartikdutt18 — CHANGES_REQUESTED** [review]
Same as the comment above.

**Comment by Aakash-kaushik** [comment]
> @zoq, Correct me if I'm wrong, but I think this should be part of activation function(https://github.com/mlpack/mlpack/tree/master/src/mlpack/methods/ann/activation_functions) not layers since it doesn't require a parameter.

There already exist a lot of activation functions in layers which are implemented as layers.

**Comment by kartikdutt18** [comment]
There is a reason for that its that they require parameters.

**Comment by Aakash-kaushik** [comment]
> There is a reason for that its that they require parameters. If a parameter isn't required it should be migrated to activation function.

Let me know if i need to do that.

**Comment by zoq** [comment]
> > There is a reason for that its that they require parameters. If a parameter isn't required it should be migrated to activation function.
> 
> Let me know if i need to do that.

We started to move out the activation functions into actual layers, the main reason is that we can have everything in one place. This will have some code duplication, but once the vtable thing is merged everything this will be simplified again.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 5]
```diff
@@ -704,6 +704,51 @@ void CheckFlattenTSwishDerivateCorrect(const arma::colvec input,
...
+}
+
+/**
+ * Basic test of the ReLU6 function.
+ */
+TEST_CASE("ReLU6FunctionTest", "[ActivationFunctionsTest]")
+{
+  const arma::colvec activationData("-2.0 3.0 0.0 6.0 24.0");
+
+  const arma::colvec desiredActivations("0.0 3.0 0.0 6.0 6.0");
```
Comment (zoq) [inline src/mlpack/tests/activation_functions_test.cpp:745]: Can you add a comment where the values are coming from, I guess from the PyTorch implementation?

Comment (Aakash-kaushik) [reply src/mlpack/tests/activation_functions_test.cpp:745]: Sure, will add a comment.

Comment (Aakash-kaushik) [reply src/mlpack/tests/activation_functions_test.cpp:746]: Done, ready for a final review.

**Review by zoq — APPROVED** [review]
Looks good to me, no further comments from my side.

**Comment by zoq** [comment]
@kartikdutt18 let us know if you like to change anything for this PR.

**Review by kartikdutt18 — APPROVED** [review]
LGTM
